File: lib/features/wallet/models/transaction_model.dart

enum TransactionType {
  credit,
  debit,
}

enum TransactionStatus {
  pending,
  successful,
  failed,
}

class Transaction {
  final String id;
  final TransactionType type;
  final double amount;
  final String description;
  final TransactionStatus status;
  final DateTime createdAt;
  final String? reference;
  
  Transaction({
    required this.id,
    required this.type,
    required this.amount,
    required this.description,
    required this.status,
    required this.createdAt,
    this.reference,
  });
  
  factory Transaction.fromJson(Map<String, dynamic> json) {
    return Transaction(
      id: json['id'],
      type: json['type'].toString().toLowerCase() == 'credit'
          ? TransactionType.credit
          : TransactionType.debit,
      amount: (json['amount'] is int)
          ? (json['amount'] as int).toDouble()
          : json['amount'],
      description: json['description'],
      status: _parseStatus(json['status']),
      createdAt: DateTime.parse(json['createdAt']),
      reference: json['reference'],
    );
  }
  
  static TransactionStatus _parseStatus(String status) {
    switch (status.toLowerCase()) {
      case 'pending':
        return TransactionStatus.pending;
      case 'successful':
        return TransactionStatus.successful;
      case 'failed':
        return TransactionStatus.failed;
      default:
        return TransactionStatus.pending;
    }
  }
  
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type == TransactionType.credit ? 'credit' : 'debit',
      'amount': amount,
      'description': description,
      'status': status.toString().split('.').last,
      'createdAt': createdAt.toIso8601String(),
      'reference': reference,
    };
  }
}

File: Dash_backend/src/models/transactionModel.js
const mongoose = require('mongoose');

const transactionSchema = new mongoose.Schema(
  {
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true,
    },
    type: {
      type: String,
      enum: ['credit', 'debit', 'commission', 'refund'],
      required: true,
    },
    amount: {
      type: Number,
      required: true,
    },
    description: {
      type: String,
      required: true,
    },
    reference: {
      type: String,
      required: true,
      unique: true,
    },
    status: {
      type: String,
      enum: ['pending', 'successful', 'failed'],
      default: 'pending',
    },
    paymentMethod: {
      type: String,
      enum: ['wallet', 'virtual_account', 'cash', 'card', 'bank_transfer'],
      required: true,
    },
    paymentDetails: {
      type: Object,
      default: {},
    },
    ride: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Ride',
      default: null,
    },
    createdAt: {
      type: Date,
      default: Date.now,
    },
    updatedAt: {
      type: Date,
      default: Date.now,
    },
  },
  {
    timestamps: true,
  }
);

const Transaction = mongoose.model('Transaction', transactionSchema);

module.exports = Transaction;
